### PR TITLE
Introduce K8sClient abstraction for easier testing

### PR DESCRIPTION
**Pull request by swoehrl-mw**
This PR introduces a new interface `K8sClient` as an abstraction over the existing `client.Client` and `ResourceReconciler`. The goals are to reduce some boilerplate code and make it easier to unittest the code with mocks instead of using the slow envtest approach everywhere.

Changes:
* New interface `K8sClient` with `K8sClientImpl` as default implementation that wraps all kubernetes client access
* Introduce [mockery](https://vektra.github.io/mockery/) as helper library to generate mocks
* Refactor all tests in `pkg/reconcilers` to use a `MockK8sClient`

With these changes it should be easier to write unittests for the reconcilers in the future. Also the runtime of the tests should be significantly reduced.

### CODE REVIEW

**Comment by saketmht** [comment]
@swoehrl-mw What is pending in this PR?

**Comment by swoehrl-mw** [comment]
> What is pending in this PR?

@saketmht 
It's waiting for a review and approval. After that I will fix the conflicts before merging (I'm not gonna fix them beforehand as there would be new ones with each other PR merged).

File: opensearch-operator/pkg/reconcilers/role_test.go
Comment (idanl21) [inline opensearch-operator/pkg/reconcilers/role_test.go:68]: Can you please explain why we deleted that step ?

File: opensearch-operator/pkg/reconcilers/scaler.go
Comment (idanl21) [inline opensearch-operator/pkg/reconcilers/scaler.go:110]: is the .UpdateOpenSearchClusterStatus is client func or our ?

File: opensearch-operator/pkg/reconcilers/suite_test.go
Comment (idanl21) [inline opensearch-operator/pkg/reconcilers/suite_test.go:78]: did we moved the build of serviceMonitorCRD to somewhere else ?

**Review by idanl21 — APPROVED** [review]

**Pull request merged**
